### ISSUE TITLE
Browser gets stuck on the starting screen

### ISSUE THREAD

**Issue by cyandotapps** [zeta0134/3ds-homebrew-browser#18]
Hello,
I downloaded the latest Homebrew to my SD card yesterday, and opened it via YouTube app exploit. Then I installed this Homebrew browser and opened it, and it gets stuck on the starting screen, and I can't do anything except turning off my 3DS XL (not new). So I can't even take a look at this browser D:

**Comment by mattycfp** [comment]
Ditto. Not working on NN3DS either.

**Comment by zeta0134** [comment]
Confirmed; this is a known bug with http:c on new versions of the homebrew exploits. Far as I'm aware, the fix is going to involve rewriting all of the networking code to utilize sockets instead, which'll be fun.

**Comment by mattycfp** [comment]
@zeta0134 It seems to have fixed itself on 0.1.1. Found out I was using the 0.1 version. However there are still no apps.

Also, if you're having trouble with hosting apps and games, I can try to set up something for ya if you'd like. I can try to get some kind of hosting.

**Comment by zeta0134** [comment]
FTP is actually considerably more complicated to support than straight HTTP. The basic issue is that right now, the homebrew browser is calling http:c, which is an OS service provided by the 3DS. It's pretty basic, only supports HTTP 1.0, and is broken in the current exploits. Sockets are better supported, and would allow the use of any regular protocol without too much fuss; it's what we should have been using all along, but http:c was easy to get working quickly at the time.

**Comment by mattycfp** [comment]
@zeta0134 Okay. Question: what would happen if you converted all of the files on the server to raw text and have the 3DS interpret that and make those into correctly named and extension'd files?

**Comment by zeta0134** [comment]
That would overcomplicate things. The act of turning network data into saved files isn't the problem, it's the transport layer that gets us that network data in the first place. Once we have the data, spitting it out to the SD card is *much* easier if it is already in the correct format.

Right now the browser basically works in reverse; the homebrew list itself is a static file, and the browser parses that and turns it into the menu client-side. All the network stack has to be able to do is download files. That's it. Everything else falls into place when that mechanism is working, and that's what's currently broken. :) (This also greatly simplifies the server setup, as literally any static file serve will do.)

The idea is of course to continue using HTTP requests. What needs to change code-wise is how those requests are made; right now we're relying on the OS to make HTTP requests on our behalf, which is broken in currently exploits. Sockets are not broken, so we simply need to make HTTP requests over sockets, and everything will magically work again. All that really takes is time; I had the thing working with sockets originally, and threw it out because the code was awful and poorly written. (*raises hand in guilt*)  Now that I have a better idea of how I want to structure things, it should be pretty easy to do. ...y'know, next week, when I'm not buried in IRL work.

tl;dr: I know how to fix the issue, it's non-trivial and will take me time, and I am tied up with IRL things until sometime next week. Fixes are incoming, you have my word. Patience please. :)

**Comment by cyandotapps** [comment]
@zeta0134 Okay ^^ As this is free, developers are allowed to take as much time as they need ^^

**Comment by mattycfp** [comment]
@zeto0134 But FTBrony does it fine. With my understanding of what you're saying, the 3DS can't download from a server. But it's well demonstrated that an FTP server works. Just use a computer server to download the actual applications then FTP them over.

**Comment by mattycfp** [comment]
@cyandotapps Have you tried updating to fix the problem?

**Comment by cyandotapps** [comment]
@mattycfp The Nintendo 3DS is actually capable of downloading things from the internet.
And yes, I tried, but this had no effect. Still the same.

**Comment by mattycfp** [comment]
I know it is capable :P

Maybe you installed it incorrectly...?

**Comment by cromo** [comment]
@mattycfp That's a whole other project on its own, largely orthogonal to the homebrew browser itself. The only thing they'd both have to know is the file layout on the server, which is presently a static file share.

I may put together a PC-side downloader at some point, but as with @zeta0134, real life takes priority, so don't bank on it. It would then be possible to FTP the downloaded files to the 3DS using other apps like FTBrony. It would be about the same amount of effort to do this than to just fix the browser in the first place, so that's what we plan to do.

**Comment by cyandotapps** [comment]
@zeta0134 After the update by Nintendo to system firmware 10.0.0-27, the Homebrew Browser (the application!) works. However, there will never be a page of downloadable applications showing up.
